### PR TITLE
refactor(optimizer): change stream key to option

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intenti    git push --set-upstream origin sts/change_stream_key_to_optionon?

a stream plan can have a stream key vec![]. So it should be None when no stream key at all.

the PR should be after https://github.com/risingwavelabs/risingwave/pull/12516

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by BugenZhao** [comment]
Related https://github.com/risingwavelabs/risingwave/pull/11718#issuecomment-1696860926

**Comment by st1page** [comment]
> Related [#11718 (comment)](https://github.com/risingwavelabs/risingwave/pull/11718#issuecomment-1696860926)

Impressed by your awesome memory :+1: :memo:

File: src/frontend/src/optimizer/plan_visitor/cardinality_visitor.rs [hunk 1]
```diff
@@ -36,11 +36,14 @@ impl CardinalityVisitor {
         input_card: Cardinality,
         eq_set: HashSet<usize>,
     ) -> Cardinality {
-        let mut unique_keys: Vec<HashSet<_>> = vec![input.stream_key().iter().copied().collect()];
-
+        let mut unique_keys: Vec<HashSet<_>> = if let Some(stream_key) = input.stream_key() {
+            vec![stream_key.iter().copied().collect()]
+        } else {
+            vec![]
+        };
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_visitor/cardinality_visitor.rs:43]: After this PR, it seems we could remove `.filter(|unique_key| !unique_key.is_empty())` this line introduced in https://github.com/risingwavelabs/risingwave/pull/11718 ?

**Review by stdrc — COMMENTED** [review]
After this, how should we interpret if we somehow get a `Some(&[])` stream key?

**Comment by st1page** [comment]
> After this, how should we interpret if we somehow get a `Some(&[])` stream key?

·`as_deref`·

File: src/frontend/src/optimizer/plan_visitor/cardinality_visitor.rs (hunk 1 above)
Comment (st1page) [reply src/frontend/src/optimizer/plan_visitor/cardinality_visitor.rs:43]: done

File: src/frontend/src/optimizer/plan_node/logical_project.rs [hunk 2]
```diff
@@ -264,7 +265,12 @@ impl ToStream for LogicalProject {
         let (proj, out_col_change) = self.rewrite_with_input(input.clone(), input_col_change);
 
         // Add missing columns of input_pk into the select list.
-        let input_pk = input.stream_key();
+        let input_pk = input.stream_key().unwrap_or_else(|| {
+            panic!(
+                "should always have a stream key in the stream plan but not, sub plan: {}",
+                input.explain_to_string()
+            )
+        });
```
Comment (stdrc) [inline src/frontend/src/optimizer/plan_node/logical_project.rs:273]: Seems to have too many same `unwrap_or_else`s, may extract to `PlanRef::expect_stream_key`?

File: src/frontend/src/optimizer/plan_node/logical_update.rs [hunk 3]
```diff
@@ -43,14 +39,8 @@ pub struct LogicalUpdate {
...
     fn from(core: generic::Update<PlanRef>) -> Self {
-        let ctx = core.input.ctx();
-        let schema = if core.returning {
-            core.input.schema().clone()
-        } else {
-            Schema::new(vec![Field::unnamed(DataType::Int64)])
-        };
-        let fd_set = FunctionalDependencySet::new(schema.len());
-        let base = PlanBase::new_logical(ctx, schema, vec![], fd_set);
+        let _ctx = core.input.ctx();
```
Comment (stdrc) [inline src/frontend/src/optimizer/plan_node/logical_update.rs:42]: This line is useless?

File: src/frontend/src/optimizer/plan_node/logical_update.rs [hunk 4]
```diff
@@ -43,14 +39,8 @@ pub struct LogicalUpdate {
 
 impl From<generic::Update<PlanRef>> for LogicalUpdate {
     fn from(core: generic::Update<PlanRef>) -> Self {
-        let ctx = core.input.ctx();
-        let schema = if core.returning {
-            core.input.schema().clone()
-        } else {
-            Schema::new(vec![Field::unnamed(DataType::Int64)])
-        };
```
Comment (stdrc) [inline src/frontend/src/optimizer/plan_node/logical_update.rs:51]: The schema is changed?

**Review by stdrc — COMMENTED** [review]
Rest LGTM

File: src/frontend/src/optimizer/plan_node/logical_update.rs (hunk 4 above)
Comment (st1page) [reply src/frontend/src/optimizer/plan_node/logical_update.rs:51]: The same logic has been implemented in generic::Update

**Review by chenzl25 — APPROVED** [review]
LGTM

**Review by BugenZhao — COMMENTED** [review]
LGTM except for https://github.com/risingwavelabs/risingwave/pull/12524/files#r1336975497

File: src/frontend/src/optimizer/plan_node/logical_project.rs (hunk 2 above)
Comment (st1page) [reply src/frontend/src/optimizer/plan_node/logical_project.rs:273]: will do it later

**Pull request merged**